src/components/TimeseriesChart/TimeseriesChart.tsx

```diff
@@ -46,13 +46,37 @@ export type TimeseriesChartProps = {
   onFetchDataError: (e: Error) => void;
 } & ({ timeseriesIds: number[] } | { series: any });
 
+// Don't allow updating faster than every 1000ms.
+const MINIMUM_UPDATE_FREQUENCY_MILLIS = 1000;
+
+export interface LegendValue {
+  id: number;
+  name: string;
+  value: number;
+  color: string;
+  timestamp: Date;
+  rawTimestamp: number;
+}
+
+export interface LegendValues {
+  [key: string]: LegendValue;
+}
+
+export interface Point {
+  timestamp?: number;
+  value?: number;
+}
+
+export interface RulerConfig {
+  timeLabel?: (point: Point) => string;
+  yLabel?: (point: Point) => string;
```

Comment:
have a look in griff-react
type used to be 
```
  export interface Ruler {
    visible?: boolean;
    timeLabel?: (point: GetAggregateDatapoint) => number;
    yLabel?: (point: GetAggregateDatapoint) => string;
    timestamp?: number;
    getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
  }

```
you are making breaking changes here

Reply:
I've been working on this all morning - the type usage is a little inconsistent here, would like some help to continue.

Let's assume `Ruler` from `griff-react` is the correct type for `ruler` prop.

```typescript
export interface Ruler {
  visible?: boolean;
  timeLabel?: (point: GetAggregateDatapoint) => number;
  yLabel?: (point: GetAggregateDatapoint) => string;
  timestamp?: number;
  getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
}
```

The only story where the `ruler` prop was used, had used it as below:

```typescript
ruler={{
  visible: true,
  yLabel: (point: any) =>
    `${Number.parseFloat(point.value).toFixed(3)}`,
  timeLabel: (point: any) => point.timestamp.toString(),
}}
```

Our assumption breaks down because above example's `yLabel` callback returns a `string` instead of a `number` as declared by the `Ruler` interface.

Also the `Ruler` interface uses `GetAggregateDatapoint` type in the timeLabel and yLabel callbacks. This is from `@cognite/sdk`

```typescript
interface GetAggregateDatapoint extends GetDatapointMetadata {
    average?: number;
    max?: number;
    min?: number;
    count?: number;
    sum?: number;
    interpolation?: number;
    stepInterpolation?: number;
    continuousVariance?: number;
    discreteVariance?: number;
    totalVariation?: number;
}

interface GetDatapointMetadata {
  timestamp: Date;
}
```

`GetAggregateDatapoint` does not have a `value` field, but the example uses `point.value`.

When I digged more in `griff-react` found below two types which actually match more closely - but they are not exported from `griff-react` or from `@cognite/sdk`

```typescript
export interface RulerConfig {
  visible: boolean;
  timeLabel: (data: RulerPoint) => string;
  yLabel: (data: RulerPoint) => string;
  getTimeLabelPosition?: (
    defaultPosition: number,
    measurements: {
      height: number;
      labelHeight: number;
      timeLabelMargin: number;
    }
  ) => number;
}

export interface RulerPoint {
  id?: ItemId;
  name: string;
  color: string;
  timestamp: number;
  value: number | string;
  x: number;
  y: number;
}
```

@polomani  / @mrtraser - need help on this please.

Reply:
The problem here is that `griff-react` doesn't provide any typings for TS. We're handling all related types by ourselves in gearbox based on internal implementation in `griff-react`. As for me, we need to:

* Fix `Ruler` interface

  ```typescript
  export interface Ruler {
    visible?: boolean;
    timeLabel?: (point: GetAggregateDatapoint) => string;
    yLabel?: (point: GetAggregateDatapoint) => string;
    timestamp?: number;
    getTimeLabelPosition?: (defaultPosition: number, params: any) => number;
  }
  ```
* Define type for `RulerConfig` same as in griff-react (just fixed) with omitted `timestamp` and `getTimeLabelPosition` props (cause we don't support them in any case)
* You don't need `RulerPoint` here at all cause you already have a `LegendValue` (witch pretty the same as `RulerPoint` above). So you need to use one type – `RulerPoint` or `LegendValue`. `RulerPoint` sounds better for me, but I would rename it to `ChartRulerPoint`

Reply:
Regarding storybook example:
```
ruler={{
  visible: true,
  yLabel: (point: any) =>
    `${Number.parseFloat(point.value).toFixed(3)}`,
  timeLabel: (point: any) => point.timestamp.toString(),
}}
```
`any` here `yLabel: (point: any) =>` and here ` timeLabel: (point: any)` actually hides the bug that we have a wrong typing in gearbox. It has to be 

```
export type DatapointsGetDatapoint =
  | DatapointsGetStringDatapoint
  | DatapointsGetDoubleDatapoint;
```
most likely (but you need to check it)

Reply:
@mrtraser could you help to figure out the type here? i should dive that deep into the code

Reply:
@polomani / @mrtraser - So I would go with `ChartRulerPoint`.

Reply:
sorry, didn't see previous message

Reply:
> @polomani / @mrtraser - So I would go with `ChartRulerPoint`.

yes, please